### ISSUE TITLE
[dev-3.x] Exponential blowup in codegen size

### ISSUE THREAD

**Issue by martinraison** [apollographql/apollo-android#3144]
**Summary**
The 3.x codegen can easily lead to huge codegen sizes, which makes compilation slow and can even break it entirely.

**Version**
3.x-dev12-SNAPSHOT

**Description**
I understand the 3.x codegen is not stable at all yet, but I thought I'd put this on your radar.

Here is the repro: https://github.com/martinraison/kmm-repro-graphql-codegen-blowup

Although the GraphQL is < 100 lines (schema + query), codegen size is > 10000 lines and grows exponentially with fragment nesting. Compilation breaks with the following error:

```
Cannot fit requested classes in a single dex file (# methods: 74868 > 65536)
```

Setting `codegenModels` to `"compat"` improves things somewhat, but `CatQuery_ResponseFields.kt` is still 8000+ lines, with a lot of class nesting, hence very long `*.class` file names which can easily exceed the filesystem limit.

Another curious symptom is that in the generated `CatQuery.kt`, `OPERATION_DOCUMENT` is extremely long. It looks like the fragments get duplicated many times.

I guess there are probably some trade-offs between ease-of-use and support for large/complex GraphQL schema, but it would be nice to have an option to avoid exponential code size blow-up. For the record, I've had this kind of issue in the past with the Apollo Client for iOS (which was solved there by setting the `mergeInFieldsFromFragmentSpreads` option to `false`), however Apollo Client for Android 2.x has never been a problem.

GraphQL schema:

```graphql
type Query {
    cat: Cat
}

type Cat {
    name: String
    field1: Cat
    field2: Cat
}
```

GraphQL query file (of course this one is a bit goofy, but I've noticed similar behavior with actual real-world business logic):

```graphql
query Cat {
    cat {
        ...CatFragment1
    }
}

fragment CatFragment1 on Cat {
    field1 {
        ...CatFragment2
    }
    field2 {
        ...CatFragment2
    }
}

fragment CatFragment2 on Cat {
    field1 {
        ...CatFragment3
    }
    field2 {
        ...CatFragment3
    }
}

fragment CatFragment3 on Cat {
    field1 {
        ...CatFragment4
    }
    field2 {
        ...CatFragment4
    }
}

fragment CatFragment4 on Cat {
    field1 {
        ...CatFragment5
    }
    field2 {
        ...CatFragment5
    }
}
fragment CatFragment5 on Cat {
    field1 {
        ...CatFragment6
    }
    field2 {
        ...CatFragment6
    }
}
fragment CatFragment6 on Cat {
    field1 {
        ...CatFragment7
    }
    field2 {
        ...CatFragment7
    }
}

fragment CatFragment7 on Cat {
    field1 {
        ...CatFragment8
    }
    field2 {
        ...CatFragment8
    }
}

fragment CatFragment8 on Cat {
    field1 {
        ...CatFragment9
    }
    field2 {
        ...CatFragment9
    }
}

fragment CatFragment9 on Cat {
    name
}
```

**Comment by martinbonnin** [comment]
First off, keep sending these, this is seriously useful ❤️ 🙏 . Thanks a lot!

> grows exponentially with fragment nesting

Right. When there are polymorphic fields, we have to generate each possible shape and that grows exponentially as nested polymorphic fields are present. There might be some optimizations like generating only once the classes with the exact same shapes but that would make class names relatively unpredictable (not to mention it's not trivial to implement) so that hasn't been done yet. We might need to explore that more.

> Setting codegenModels to "compat" improves things somewhat, but CatQuery_ResponseFields.kt is still 8000+ lines, 

I'll compare with 2.x. It should be the ~same length. If not, it's a bug.

> with a lot of class nesting, hence very long *.class file names which can easily exceed the filesystem limit.

The latest snapshot should have a `flattenModels` option to flatten the class hierarchy (with the tradeoff that there will potentially be classes with the same names being suffixed like in 2.x)

> Another curious symptom is that in the generated CatQuery.kt, OPERATION_DOCUMENT is extremely long. It looks like the fragments get duplicated many times.

Sounds like a bug, will look into this.

> it would be nice to have an option to avoid exponential code size blow-up

`codegenModels.set("compat")` (or even better `codegenModels.set("responseBased")`) should be O(n) where n is the number of fields queried. 

> I've noticed similar behavior with actual real-world business logic

What level nesting of polymorphic fields do you have in your queries? The highest I can remember is ~2, maybe 3. Are the fields actually polymorphic or do they just happen to be polymorphic because they are interfaces but you only query one type ? Maybe we could do something for these and do not generate the fallback type or something like this...

I'll put together a high level document to go through the different options. It's something I meant to do for a while now. Will post the PR here.

**Comment by martinraison** [comment]
Thanks for all the context ! I'm doing some early evaluation of Kotlin Multiplatform on a relatively large codebase so I'll keep sending issues as I bump into them :)

> `codegenModels.set("compat")` (or even better `codegenModels.set("responseBased")`) should be O(n) where n is the number of fields queried.

With that option enabled, all generated files are small except for:

- `CatQuery.kt` but only because `OPERATION_DOCUMENT` is huge (which I understand is a bug)
- `CatQuery_ResponseFields.kt`, which does seem to grow exponentially in size. If I remove one level of nesting I go from 8k lines to 4k lines. And if I add one level of nesting I get this curious error (which I didn't notice before):

```
> Task :shared:compileDebugKotlinAndroid
e: org.jetbrains.kotlin.backend.common.BackendException: Backend Internal error: Exception during code generation
File being compiled: /Users/martinraison/sandbox/kmm-repro-graphql-codegen-blowup/shared/build/generated/source/apollo/service/com/example/repro/graphql/CatQuery.kt
The root cause java.lang.IllegalArgumentException was thrown at: org.jetbrains.org.objectweb.asm.ByteVector.putUTF8(ByteVector.java:246)
	at org.jetbrains.kotlin.backend.common.CodegenUtil.reportBackendException(CodegenUtil.kt:239)
	at org.jetbrains.kotlin.backend.common.CodegenUtil.reportBackendException$default(CodegenUtil.kt:235)
[...]
Caused by: java.lang.IllegalArgumentException: UTF8 string too large
	at org.jetbrains.org.objectweb.asm.ByteVector.putUTF8(ByteVector.java:246)
	at org.jetbrains.org.objectweb.asm.SymbolTable.addConstantUtf8(SymbolTable.java:774)
	at org.jetbrains.org.objectweb.asm.SymbolTable.addConstantUtf8Reference(SymbolTable.java:1007)
	at org.jetbrains.org.objectweb.asm.SymbolTable.addConstantString(SymbolTable.java:604)
	at org.jetbrains.org.objectweb.asm.SymbolTable.addConstant(SymbolTable.java:474)
	at org.jetbrains.org.objectweb.asm.FieldWriter.<init>(FieldWriter.java:136)
	at org.jetbrains.org.objectweb.asm.ClassWriter.visitField(ClassWriter.java:436)
	at org.jetbrains.kotlin.codegen.AbstractClassBuilder.newField(AbstractClassBuilder.java:70)
	at org.jetbrains.kotlin.codegen.DelegatingClassBuilder.newField(DelegatingClassBuilder.java:41)
	at org.jetbrains.kotlin.codegen.DelegatingClassBuilder.newField(DelegatingClassBuilder.java:41)
	at org.jetbrains.kotlin.backend.jvm.codegen.ClassCodegen.generateField(ClassCodegen.kt:297)
	at org.jetbrains.kotlin.backend.jvm.codegen.ClassCodegen.generate(ClassCodegen.kt:139)
	at org.jetbrains.kotlin.backend.jvm.JvmIrCodegenFactory.doGenerateFilesInternal(JvmIrCodegenFactory.kt:206)
	... 34 more

> Task :shared:compileDebugKotlinAndroid FAILED
```

> What level nesting of polymorphic fields do you have in your queries? The highest I can remember is ~2, maybe 3. Are the fields actually polymorphic or do they just happen to be polymorphic because they are interfaces but you only query one type ? Maybe we could do something for these and do not generate the fallback type or something like this...

We go to at least 5 (whether that's a good thing is another question, but that's we have right now :) ). In fact, the first hard limit I bumped into was that `*.class` names were too long (because we also have a good amount of non-polymorphic nesting, and relatively long field names).

The same setup compiles correctly with 2.x.

> I'll put together a high level document to go through the different options. It's something I meant to do for a while now. Will post the PR here.

Thanks, sounds great !

**Comment by martinbonnin** [comment]
> because OPERATION_DOCUMENT is huge (which I understand is a bug)

Silly bug confirmed! The fix is there: https://github.com/apollographql/apollo-android/pull/3145

> CatQuery_ResponseFields.kt, which does seem to grow exponentially in size. 

That's the code responsible for reading the queries from the cache, makes sense to not want that exponential indeed, will look into it.

> We go to at least 5 

 😮  Got it! "compat" or "operationBased" with `flattenModels.set(true)` is most likely what you want then. "operationBased" will even be slightly smaller than "compat". I tend to like it better as it maps 1:1 with the queries but "compat" might feel more familiar. 
The repo is moving fast at the moment. I'll look into the ResponseFields and update this ticket. Thanks again for the feedback!

**Comment by martinraison** [comment]
> Silly bug confirmed! The fix is there: #3145

Awesome, that was a quick fix!

> That's the code responsible for reading the queries from the cache, makes sense to not want that exponential indeed, will look into it.

Great, happy to help testing if needed

> 😮 Got it! "compat" or "operationBased" with `flattenModels.set(true)` is most likely what you want then. "operationBased" will even be slightly smaller than "compat". I tend to like it better as it maps 1:1 with the queries but "compat" might feel more familiar.

I haven't tested `operationBased` yet but it sounds useful, I'll give it a try. Thanks for the suggestion.

**Comment by martinbonnin** [comment]
First drop for a Codegen doc: https://github.com/apollographql/apollo-android/blob/f2267c41cbe3863673d009b919d590904f7e0412/design-docs/Codegen.md (from #3149)

And the PR to have O(fields) cache helpers: https://github.com/apollographql/apollo-android/pull/3148

Overall, the "responseBased" codegen can lead to huge generated classes. Looking at your example again, there's actually no polymorphism so it's more about reusing the fragments rather making them available as interfaces all the time. I'll try to think about this. 

In the short term, "compat" or "operationBased" is certainly the safest bet. If the features of "compat" aren't actually needed, we could rename it to "default". Even if I'd prefer to keep only 2 options in the long run.

**Comment by martinraison** [comment]
Many thanks for the PR and docs !

The doc is very useful, everything is much clearer now. It looks like operation-based models is what we're looking for indeed.

I hadn't considered the performance implications, but it hasn't really been an issue for us so far. The network response buffering is a fun question though: from my superficial understanding of the problem, it seems we could avoid having to buffer the entire response in a HashMap, as long as the server behaves nicely and returns the `__typename` field first (so the client can immediately decide which fragment type to start populating). Of course, the library would still need to support other server behaviors, so it would increase the implementation complexity.

I'll test the PR as soon as a new snapshot is available (the latest build failed)

**Comment by martinraison** [comment]
I can confirm the problem is fixed with the latest build. Thanks again !

**Comment by martinbonnin** [comment]
I'm closing this issue for now. We can try to think of ways to mix the type information of response-based codegen with the minimal codegen size of the operation-based one but that's more longer term. 
Do not hesitate to leave a message or open another issue if you feel there's something that still needs to be changed.
